runner/scheduler.py

```diff
@@ -254,29 +247,35 @@ def pick_next_test_id(self, tests: dict[str, TestExample]) -> str:
 
         assert False, f"Couldn't find a test to run. Wait list: {self.wait_list}"
 
-    def iter_tests(self, test_group: list[TestExample]) -> Iterator[TestExample]:
-        # Setup: get the list of tests that are in progress
+    def setup_iterator(self, test_group):
+        # Sets up the test dict and fast forwards any tests that are currently in progress
         test_actions_taken = self.master_log.get_tests_in_progress()
-        # TODO: More detailed assemblage of wait list
         self.wait_list = {k: {"tokens": 0} for k in test_actions_taken.keys()}
         tests = {t.unique_id: t for t in test_group}
 
+        # Check if the last event in the log is after the current time, then travel to that time if it is.
+        if len(self.master_log.log) > 0 and datetime.now() < self.master_log.log[-1].timestamp:
+            self.travel_to_dt(self.master_log.log[-1].timestamp)
+
         # Fast forward examples to the last action that was run
-        for run_id in self.wait_list.keys():
+        for run_id in test_actions_taken.keys():
             example = tests[run_id]
             actions_to_ff = test_actions_taken[run_id]
             for _ in range(actions_to_ff):
                 action = example.step()
                 if isinstance(action, SendAndRegisterAction):
                     self.register_callback(example)
 
-                if isinstance(action, WaitAction) and action.time.seconds > 0:
-                    self.forward_time(seconds=action.time.seconds + 1)
-
             # If the last action is a wait action, set the test to wait
-            if isinstance(action, WaitAction):
+            if isinstance(action, WaitAction) and action.tokens > 0:
                 self.set_to_wait(run_id, action)
```

Comment:
I'm not sure about this, but it might be important that the test is set to wait also for time waits, because waiting tests are given priority in the scheduling.

Reply:
Good point, the wait could be logged and the run die before the wait actually concludes.